File: packages/validator-rust/src/legacy/spec.rs
//! Spec parsing: canonical form-spec JSON (type/properties/rules) -> Spec.
//! Ports validator-go/validator/spec.go. Property and rule declaration order
//! is preserved because serde_json is built with the preserve_order feature,
//! so Map iteration follows insertion order.

use crate::legacy::types::{CustomRule, Field, Spec};
use serde_json::{Map, Value};

/// ParsedSpec is the result of parsing a canonical form-spec document.
/// is_group reports whether the root was a group spec with properties;
/// non-group specs are wrapped into a single field named "value".
pub struct ParsedSpec {
    pub spec: Spec,
    pub is_group: bool,
}

/// parse_spec parses a canonical form-spec JSON value into a Spec.
pub fn parse_spec(root: &Value) -> ParsedSpec {
    let obj = match root.as_object() {
        Some(o) => o,
        None => {
            return ParsedSpec {
                spec: Spec::default(),
                is_group: false,
            }
        }
    };

    let type_str = obj.get("type").and_then(Value::as_str).unwrap_or("");
    if type_str == "group" {
        if let Some(props) = obj.get("properties").and_then(Value::as_object) {
            return ParsedSpec {
                spec: Spec {
                    fields: convert_properties(props),
                    rules: Default::default(),
                },
                is_group: true,
            };
        }
    }

    // Wrap simple field spec in a group with a 'value' property.
    ParsedSpec {
        spec: Spec {
            fields: vec![convert_field("value", obj)],
            rules: Default::default(),
        },
        is_group: false,
    }
}

/// convert_properties converts a properties object to fields in declaration
/// order.
fn convert_properties(props: &Map<String, Value>) -> Vec<Field> {
    let mut fields = Vec::with_capacity(props.len());
    for (name, raw) in props {
        if let Some(field_obj) = raw.as_object() {
            fields.push(convert_field(name, field_obj));
        }
    }
    fields
}

/// convert_field converts a single field spec object to a Field.
fn convert_field(name: &str, obj: &Map<String, Value>) -> Field {
    let mut field = Field {
        name: name.to_string(),
        ..Default::default()
    };

    field.field_type = obj.get("type").and_then(Value::as_str).unwrap_or("").to_string();
    field.label = obj.get("label").and_then(Value::as_str).unwrap_or("").to_string();

    if let Some(rules_obj) = obj.get("rules").and_then(Value::as_object) {
        for (rule_name, rule_val) in rules_obj {
            field.rules.insert(rule_name.clone(), rule_val.clone());
            field.rule_order.push(rule_name.clone());
        }
    }

    if let Some(msg_obj) = obj.get("messages").and_then(Value::as_object) {
        for (msg_key, msg_val) in msg_obj {
            if let Some(s) = msg_val.as_str() {
                if !s.is_empty() {
                    field.messages.insert(msg_key.clone(), s.to_string());
                }
            }
        }
    }

    if let Some(props_obj) = obj.get("properties").and_then(Value::as_object) {
        field.fields = convert_properties(props_obj);
    }

    if let Some(mv) = obj.get("multiple") {
        match mv {
            Value::Bool(b) => field.multiple = *b,
            Value::String(s) if s == "only" => field.multiple_only = true,
            _ => {}
        }
    }

    if let Some(ds) = obj.get("display_switch") {
        field.display_switch = Some(ds.clone());
    }

    if let Some(dt) = obj.get("display_target").and_then(Value::as_str) {
        if !dt.is_empty() {
            field.display_target = dt.to_string();
        }
    }

    // Legacy top-level required attribute (bool or condition string).
    if let Some(req) = obj.get("required") {
        if !req.is_null() {
            field.required = Some(req.clone());
        }
    }

    field
}

/// build_custom_rules parses spec-level named rules (optional; the shared
/// fixtures define rules inside fields, but this mirrors the Go Spec.Rules
/// channel).
#[allow(dead_code)]
pub fn build_custom_rules(root: &Value) -> std::collections::HashMap<String, CustomRule> {
    let mut out = std::collections::HashMap::new();
    if let Some(rules_obj) = root.get("rules").and_then(Value::as_object) {
        for (name, raw) in rules_obj {
            if let Some(obj) = raw.as_object() {
                let rule = CustomRule {
                    pattern: obj.get("pattern").and_then(Value::as_str).map(str::to_string),
                    min: obj.get("min").and_then(Value::as_i64),
                    max: obj.get("max").and_then(Value::as_i64),
                    message: obj.get("message").and_then(Value::as_str).unwrap_or("").to_string(),
                };
                out.insert(name.clone(), rule);
            }
        }
    }
    out
}

File: packages/validator-go/validator/legacy/spec.go
package legacy

import "strings"

// ParsedSpec is the result of parsing a canonical crudui JSON document.
// IsGroup reports whether the document root was a group spec with properties;
// non-group specs are wrapped into a single field named "value".
type ParsedSpec struct {
	Spec    Spec
	IsGroup bool
}

// ParseSpec parses a canonical crudui JSON document
// (type/properties/rules format) into a validator Spec.
// Property and rule declaration order is preserved.
func ParseSpec(data []byte) (ParsedSpec, error) {
	decoded, err := DecodeOrderedJSON(data)
	if err != nil {
		return ParsedSpec{}, err
	}

	om, ok := decoded.(*OrderedMap)
	if !ok {
		return ParsedSpec{Spec: Spec{}}, nil
	}

	props, hasProps := om.Get("properties")
	propsOM, propsIsMap := props.(*OrderedMap)
	if om.GetString("type") == "group" && hasProps && propsIsMap {
		return ParsedSpec{
			Spec:    Spec{Fields: convertOrderedProperties(propsOM)},
			IsGroup: true,
		}, nil
	}

	// Wrap simple field spec in a group with a 'value' property
	return ParsedSpec{
		Spec:    Spec{Fields: []Field{convertOrderedField("value", om)}},
		IsGroup: false,
	}, nil
}

// convertOrderedProperties converts a properties object to fields in
// declaration order.
func convertOrderedProperties(props *OrderedMap) []Field {
	fields := make([]Field, 0, props.Len())
	for _, name := range props.Keys() {
		raw, _ := props.Get(name)
		if fieldOM, ok := raw.(*OrderedMap); ok {
			fields = append(fields, convertOrderedField(name, fieldOM))
		}
	}
	return fields
}

// convertOrderedField converts a single field spec object to a Field.
//
// A "[]"-suffix property key (e.g. "items[]") is legacy Limepie shorthand for
// a multiple field: the suffix is stripped from the field name (so it binds to
// the "items" data key) and Multiple is forced on. Without this the raw
// "items[]" key never matches the data and validation is silently skipped
// (PHP Validator.php and client dist.validate.js parity).
func convertOrderedField(name string, om *OrderedMap) Field {
	arraySuffix := strings.HasSuffix(name, "[]")
	if arraySuffix {
		name = strings.TrimSuffix(name, "[]")
	}

	field := Field{Name: name}
	if arraySuffix {
		field.Multiple = true
	}

	field.Type = om.GetString("type")
	field.Label = om.GetString("label")

	if raw, ok := om.Get("rules"); ok {
		if rulesOM, ok := raw.(*OrderedMap); ok {
			field.Rules = make(map[string]interface{}, rulesOM.Len())
			field.RuleOrder = make([]string, 0, rulesOM.Len())
			for _, ruleName := range rulesOM.Keys() {
				ruleVal, _ := rulesOM.Get(ruleName)
				field.Rules[ruleName] = ToPlain(ruleVal)
				field.RuleOrder = append(field.RuleOrder, ruleName)
			}
		}
	}

	if raw, ok := om.Get("messages"); ok {
		if msgOM, ok := raw.(*OrderedMap); ok {
			field.Messages = make(map[string]string, msgOM.Len())
			for _, msgKey := range msgOM.Keys() {
				if s := msgOM.GetString(msgKey); s != "" {
					field.Messages[msgKey] = s
				}
			}
		}
	}

	if raw, ok := om.Get("properties"); ok {
		if propsOM, ok := raw.(*OrderedMap); ok {
			field.Fields = convertOrderedProperties(propsOM)
		}
	}

	if raw, ok := om.Get("multiple"); ok {
		switch mv := raw.(type) {
		case bool:
			// A "[]"-suffix key already forced Multiple on; an explicit
			// multiple key can only widen it, never turn it off.
			field.Multiple = field.Multiple || mv
		case string:
			if mv == "only" {
				field.MultipleOnly = true
			}
		}
	}

	if raw, ok := om.Get("display_switch"); ok {
		field.DisplaySwitch = ToPlain(raw)
	}

	if dt := om.GetString("display_target"); dt != "" {
		field.DisplayTarget = dt
	}

	return field
}
